Implemented BccProgress component

steffanhalv opened the pull request:
# Change summary

## Change type

-   [ ] No review
-   [x] Small PR
-   [ ] Big PR
-   [ ] Refactor

Fixes #22
Fixes #238

steffanhalv committed:
Implemented BccProgress component

laurensgroeneveld commented:
I see there's quite a bit of linting errors and realized there was not yet anything about it in the readme. I've added a section on linting now: https://github.com/bcc-code/bcc-design/pull/246

steffanhalv committed:
Fix linting

steffanhalv committed:
Merge branch 'main' into feature/BccProgress

steffanhalv commented:
> I see there's quite a bit of linting errors and realized there was not yet anything about it in the readme. I've added a section on linting now: #246

Fixed @laurensgroeneveld

steffanhalv committed:
Fix typecheck

laurensgroeneveld commented on design-library/src/components/BccProgress/BccProgress.stories.ts:
```diff
@@ -0,0 +1,129 @@
+ * A `progress` element supporting current and max values
+ */
+export default {
+  title: "Forms/BccProgress",
```
I think the Forms section should be reserved for elements that you'd use in an HTML form, like inputs and buttons. I'll admit this one is a bit on the edge but I suggest placing it in Components for now (we can create more subcategories as the component list grows).

laurensgroeneveld commented on design-library/src/components/BccProgress/BccProgress.vue:
```diff
@@ -0,0 +1,74 @@
+  <div
+    class="bcc-progress-container"
+    :class="$attrs['class']"
+    :style="$attrs['style'] as StyleValue"
```
Vue will add any `class` and `style` elements by default on the outer element, so if that's the intention then this is not needed. In some cases, like the Input component, it seemed to make more sense to add classes to the `input` element itself, not the wrapping `div`. But in this case I think it's fine to put the classes on here, and then this code is not needed.

laurensgroeneveld commented on design-library/src/components/BccProgress/BccProgress.vue:
```diff
@@ -0,0 +1,74 @@
+export default {
+  inheritAttrs: false,
+};
+</script>
```
See above comment.

laurensgroeneveld commented on design-library/src/components/BccProgress/BccProgress.vue:
```diff
@@ -0,0 +1,74 @@
+  return {
+    "h-1": props.size === "sm",
+    "h-1.5": props.size === "base",
+    "h-2": props.size === "lg",
```
These can be custom classes as well, `bcc-progress-sm` etc. (just to make it easier for people using the library as a CSS library).

laurensgroeneveld reviewed (changes requested):
Really nice component with great customization 👌 Just got a few small comments, let me know if you don't agree with something!

laurensgroeneveld commented:
Hmm I see the Storybook deploy is failing because there is no deploy token. Unfortunately there doesn't seem to be a way other than [changing the workflow trigger](https://securitylab.github.com/research/github-actions-preventing-pwn-requests/) to change this, but that has a lot of downsides so I don't think that's a good idea.

Since you're a member of bcc-code, it should work if the branch is created on the repository (instead of a fork). I've tested it locally though so for this PR it's fine, I'll bypass the failing build when merging 👍

steffanhalv committed:
Add full customization

steffanhalv commented:
I did some other updates now to allow full customization, moved it to components and removed the $attrs inheritation with a story for how to fully customize the component.

@laurensgroeneveld

laurensgroeneveld commented on design-library/src/components/BccProgress/BccProgress.vue:
```diff
@@ -36,37 +33,52 @@ const percentageSafe = computed(() => {
+    };
+  }
+  return progressClass.value;
+});
```
I'm a bit ambivalent towards this. On the one hand it's a nice way of handing control to the consumer of the library. On the other hand, an explicit goal of this library is to have a limited set of options, so developers are forced to work within the constraints of the design system (perhaps this goal should be better articulated somewhere). I personally feel this makes the code a bit less readable as well.

I'd be inclined to suggest to not add these class props. Instead, people can override the classes that are already there (like `bcc-progress`) in the normal CSS way. The Event app (my.event.bcc.no) does this to add dark mode support for example.

What do you think, do you agree with the above reasoning? Maybe we should have a broader discussion about this?

steffanhalv replied:
Agreed, simplified it now 👍🏼

steffanhalv committed:
Simplify progress bar

steffanhalv commented:
Check update @laurensgroeneveld

steffanhalv committed:
Simplified css

steffanhalv committed:
Removed whitespace

steffanhalv committed:
Removed comment header

laurensgroeneveld reviewed (approved):
Nice work 🙌

steffanhalv committed:
Merge branch 'main' into feature/BccProgress
